src/pudl_archiver/archivers/classes.py

```diff
@@ -364,6 +367,73 @@ def _check_dataset_size(
             required_for_run_success=self.fail_on_dataset_size_change,
         )
 
+    def _check_data_continuity(
+        self, new_datapackage: DataPackage
+    ) -> validate.DatasetSpecificValidation:
+        """Check that the archived data partitions are continuous and unique."""
+        success = True
+        note = None
+        partition_names = []
+        dataset_partitions = []
+
+        # Unpack partitions of a dataset.
+        for resource in new_datapackage.resources:
+            if not resource.parts:  # Skip resources without partitions
+                continue
+            partition_names += list(resource.parts)
+            dataset_partitions += list(resource.parts.values())[0]
+
+        partition_names = set(partition_names)
+
+        # Only perform this test if the part label is year quarter or year month
+        # Note that this currently only works if there is one set of partitions,
+        # and will fail if year_month and form are used to partition a dataset, e.g.
+        if any(partition in VALID_PARTITION_RANGES for partition in partition_names):
+            partition_to_test = [
+                partition
+                for partition in partition_names
+                if partition in VALID_PARTITION_RANGES
+            ][0]
+            interval = (
+                "QS"
+                if partition_to_test == "year_quarter"
+                else "MS"
+                if partition_to_test == "year_month"
+                else None
+            )
+
+            if interval:
+                expected_date_range = pd.date_range(
+                    min(dataset_partitions), max(dataset_partitions), freq=interval
+                )
+                observed_date_range = pd.to_datetime(dataset_partitions)
+
+                if observed_date_range.has_duplicates:
+                    success = False
+                    note = [
+                        f"Partition contains duplicate time periods of data: f{observed_date_range.duplicated()}"
+                    ]
+
+                else:
+                    diff = expected_date_range.difference(observed_date_range)
+                    if not diff.empty:
+                        success = False
+                        note = [
+                            f"Downloaded partitions are not consecutive. Missing the following {partition_to_test} partitions: {diff.to_numpy()}"
+                        ]
```

Comment:
i think this would be more readable if this was a combined `elif`

```suggestion
                elif not diff.empty:
                    success = False
                    note = [
                            f"Downloaded partitions are not consecutive. Missing the following {partition_to_test} partitions: {diff.to_numpy()}"
                        ]
                else:
                    success = True
                    note = None
```

Reply:
Agreed and added!